fe/fe-core/src/main/java/com/starrocks/external/hive/HiveMetaClient.java

```diff
@@ -202,6 +212,67 @@ public HivePartition getPartition(String dbName, String tableName, List<String>
+                Partition partition = client.hiveClient.getPartition(dbName, tableName, partitionValues);
+                sd = partition.getSd();
+                partName = FSUtils.getRelativePartitionPath(new Path(basePath), new Path(sd.getLocation()));
+            }
+            HdfsFileFormat format = HdfsFileFormat.fromHdfsInputFormatClass(sd.getInputFormat());
+            if (format == null) {
+                throw new DdlException("unsupported file format [" + sd.getInputFormat() + "]");
+            }
+            String path = ObejctStorageUtils.formatObjectStoragePath(sd.getLocation());
+            List<HdfsFileDesc> fileDescs = getHudiFileDescs(sd, basePath, partName);
+            return new HivePartition(format, ImmutableList.copyOf(fileDescs), path);
+        } catch (NoSuchObjectException e) {
+            throw new DdlException("get hudi partition meta data failed: "
+                    + "partition not exists, partitionValues: "
+                    + String.join(",", partitionValues));
+        } catch (Exception e) {
+            LOG.warn("get partition failed", e);
+            throw new DdlException("get hudi partition meta data failed: " + e.getMessage());
+        }
+    }
+
+    private List<HdfsFileDesc> getHudiFileDescs(StorageDescriptor sd, String basePath, String partName) throws Exception {
+        List<HdfsFileDesc> fileDescs = Lists.newArrayList();
+
+        FileSystem fileSystem = getFileSystem(new URI(basePath));
+        // get properties like 'field.delim' and 'line.delim' from StorageDescriptor
+        TextFileFormatDesc textFileFormatDesc = new TextFileFormatDesc(
+                sd.getSerdeInfo().getParameters().getOrDefault("field.delim", ","),
+                sd.getSerdeInfo().getParameters().getOrDefault("line.delim", "\n"));
+
+        Configuration conf = new Configuration();
+        HoodieTableMetaClient metaClient = HoodieTableMetaClient.builder().setConf(conf).setBasePath(basePath).build();
+        HoodieEngineContext engineContext = new HoodieLocalEngineContext(conf);
+        HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder().build();
+        HoodieTableFileSystemView fileSystemView = FileSystemViewManager.createInMemoryFileSystemView(engineContext,
+                metaClient, metadataConfig);
+        Iterator<HoodieBaseFile> hoodieBaseFileIterator = fileSystemView.getLatestBaseFiles(partName).iterator();
+        while (hoodieBaseFileIterator.hasNext()) {
+            HoodieBaseFile baseFile = hoodieBaseFileIterator.next();
+
+            FileStatus fileStatus = HoodieInputFormatUtils.getFileStatus(baseFile);
+            BlockLocation[] blockLocations;
+            if (fileStatus instanceof LocatedFileStatus) {
+                blockLocations = ((LocatedFileStatus) fileStatus).getBlockLocations();
+            } else {
+                blockLocations = fileSystem.getFileBlockLocations(fileStatus, 0, fileStatus.getLen());
```

Comment:
If the code goes here, getting block locations for each file will be very slow when there are too many files.

Reply:
yes, Trino/Presto uses thread pool to speed up here and it may need more works to do here. I will Investigate the handling of other engines like spark to improve the process later.